Mobile workers list: angular => knockout

orangejenny opened the pull request:
@Rohit25negi 
or @proteusvacuum since you're good at knockout

Product: the user-facing changes here are minor updates to the pagination controls.

Before (doctored to show 5 workers instead of 10)
<img width="1132" alt="screen shot 2018-09-27 at 9 57 46 pm" src="https://user-images.githubusercontent.com/1486591/46183759-6d5a7200-c2a0-11e8-887d-981a5fd9bf6a.png">

After
<img width="1128" alt="screen shot 2018-09-27 at 9 55 23 pm" src="https://user-images.githubusercontent.com/1486591/46183715-2a989a00-c2a0-11e8-83c9-baa2f63320f5.png">

orangejenny committed:
Comment out markup for list, which will be turned to knockout

orangejenny committed:
Fetch first page of users and display basic info

orangejenny committed:
Split pagination out as separate view

orangejenny committed:
Set up spinner, error, and no users messages

orangejenny committed:
Support search box

orangejenny committed:
Support active/deactivated toggle; kill top pagination controls

orangejenny committed:
Support multiple pages

orangejenny committed:
Support activate/deactivate buttons

orangejenny committed:
Delete no-longer-used angular code

orangejenny committed:
Remove now-defunct include_location option

Added in 078f66cfd6703b7b9bed8cfa53c1f886a73fa22f andthen removed in 0735008a60c335cf887c502119f1f69dd4d36da5

orangejenny committed:
Remove custom user data

It's used when adding a new user, but not used for listing existing workers.

orangejenny committed:
lint

proteusvacuum commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -0,0 +1,121 @@
+                    page: page || 1,
+                    query: self.query(),
+                    limit: self.itemsPerPage(),
+                    showDeactivatedUsers: self.inactiveOnly() ? 1 : 0,
```
I think you can pass the boolean value directly (`self.inactiveOnly`) then do `json.loads` on the backend, so you don't have to do all of these conversion steps.

orangejenny replied:
a00971a

proteusvacuum commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -0,0 +1,121 @@
+    var mobileWorker = function (options) {
+        var self = ko.mapping.fromJS(options);
+
+        self.mark_activated = ko.observable(false);
```
any reason you aren't using the camelCase convention here?

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+            'username': user.raw_username,
+            'first_name': user.first_name,
+            'last_name': user.last_name,
+            'phoneNumbers': user.phone_numbers,
```
this isn't used anywhere in the frontend AFAICT

orangejenny replied:
5b03b86

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+            'mark_deactivated': False,
+            'dateRegistered': user.created_on.strftime(USER_DATE_FORMAT) if user.created_on else '',
+            'editUrl': reverse(EditCommCareUserView.urlname, args=[domain, user.user_id]),
+            'deactivateUrl': "#",
```
not used

orangejenny replied:
5b03b86

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+            'dateRegistered': user.created_on.strftime(USER_DATE_FORMAT) if user.created_on else '',
+            'editUrl': reverse(EditCommCareUserView.urlname, args=[domain, user.user_id]),
+            'deactivateUrl': "#",
+            'actionText': _("Deactivate") if user.is_active else _("Activate"),
```
not used

orangejenny replied:
5b03b86

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+            'last_name': user.last_name,
+            'phoneNumbers': user.phone_numbers,
+            'user_id': user.user_id,
+            'mark_activated': False,
```
I don't think you need to send these along like this since they are already defaulted to false in the ko model.

orangejenny replied:
5b03b86

proteusvacuum commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -0,0 +1,121 @@
+        self.mark_deactivated = ko.observable(false);
+        self.action_error = ko.observable('');
+
+        self.showActivate = ko.computed(function () {
```
I think you could remove a lot of this logic.
1. don't send `action` from the backend (instead, I would send `status` or more preferably `is_active`, so it matches the django model)
2. don't compute showActivate or showDeactivate
3. in the template, instead of `visible: showActivate`, it might be clearer to do `status === deactivated` or `!isActive`. That way the action is clearly linked to the status of the mobile worker. 
4. Then, in your `modifyStatus` function, you don't need to have the cryptic boolean argument (since `isActive` will be part of the view model). 

you might also then be able to remove the `mark_deactivated` and `mark_activated` fields, but I think I might be missing the nuance of why those exist.

proteusvacuum commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -0,0 +1,121 @@
+            return !self.notLoaded() && !self.hasError() && !self.showNoUsers() && !self.showProjectHasNoUsers();
+        });
+
+        self.toggleInactiveOnly = function (inactiveOnly) {
```
this function doesn't actually toggle the value, but sets it to the value of the argument. I'd call it `showInactiveOnly` or something similar. 
Alternatively, in the template, you could directly set the property of `inactiveOnly`, and set up a listener for a change to that value which calls `goToPage(1)`, i.e.: 
```js
self.inactiveOnly.subscribe(function () { self.goToPage(1); });
```

orangejenny replied:
a00971a

proteusvacuum commented on corehq/apps/users/templates/users/mobile_workers.html:
```diff
@@ -242,54 +240,29 @@ <h3 class="panel-title">{% trans 'Mobile Workers' %}</h3>
+                <div class="alert alert-info" data-bind="visible: showProjectHasNoUsers">
                     {% blocktrans %}
                         <strong>There are currently no Mobile Workers in this project.</strong>
                         Please Create a new Mobile Worker above to get started.
```
nit, "Create" and "Mobile Workers" should be lower case

orangejenny replied:
7143e8b

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+    # backend pages start at 0
+    users_query = _user_query(query, page - 1, limit)
+    # run with a blank query to fetch total records with same scope as in search
+    total_records = _user_query('', 0, 0).count()
```
I don't think you show this value any more

orangejenny replied:
d659a0e

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+        'users': [_format_user(user) for user in users_data.hits],
+        'total': users_data.total,
+        'page': page,
+        'query': query,
```
you don't need to send the page or the query back, because I don't think you use them

proteusvacuum commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+        users_query = users_query.show_only_inactive()
+    users_data = users_query.run()
+    return json_response({
+        'users': [_format_user(user) for user in users_data.hits],
```
so I think, since you are only using a few fields from the user es model, you could actually do: 
```py
users_data = users_query.source(['first_name', 'last_name', 'raw_username', '_id', 'created_on']).run()
``` 
and it will pre-format it for you and only pull those values from ES (i.e. it needs to pull less data from ES)

orangejenny replied:
Because `_format_user` calls `wrap`, I need to either stop wrapping (and duplicate `raw_username` logic) or include the various fields needed to wrap, and both of those options are messier than the current code.

proteusvacuum replied:
`base_username` in elasticsearch is `raw_username` on the wrapped document (https://github.com/dimagi/commcare-hq/blob/465c6380133d291df51b9433da90ea9d4c40a832/corehq/pillows/user.py#L49-L52) Why? Who knows.
I'm saying  you can just remove the whole `_format_user` function if you pull out just the fields that you want from ES, and use its output in your view. This means you don't have to wrap each user, and you only retrieve the amount of data you actually are using from ES. 
```py
In [14]: es.UserES().mobile_users().fields(['first_name', 'last_name', 'base_username', '_id', 'created_on', 'is_active']).run().hits[5]
Out[14]: 
{u'_id': u'1ef76033c5a64b659fac741ac4fb2c62',
 u'base_username': u'test2',
 u'created_on': u'2018-10-03T18:18:59.194435Z',
 u'first_name': u'',
 u'is_active': True,
 u'last_name': u''}
```

orangejenny replied:
Aha, knowing about `base_username` was the missing link. 91ee416

proteusvacuum reviewed (commented):
I'm into it! Way simpler than before!

orangejenny committed:
Merge branch 'master' into jls/ko-mobile-workers-list-redo

orangejenny committed:
Fix test

orangejenny committed:
Move editUrl to front end

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+            'mark_activated': False,
+            'mark_deactivated': False,
+            'dateRegistered': user.created_on.strftime(USER_DATE_FORMAT) if user.created_on else '',
+            'editUrl': reverse(EditCommCareUserView.urlname, args=[domain, user.user_id]),
```
63e9b2d (I agree about not passing urls like this, was just being lazy about how much code I changed.)

orangejenny committed:
clearer naming: notLoaded => isLoaded

orangejenny commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -0,0 +1,121 @@
+
+        // Visibility of spinner, messages, and user table
+        self.hasError = ko.observable(false);
+        self.notLoaded = ko.observable(true);
```
354e7db

orangejenny committed:
Remove unused fields and make casing more consistent

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+
+    def _format_user(user_json):
+        user = CouchUser.wrap_correctly(user_json)
+        return {
```
5b03b86

orangejenny committed:
Fix casing

proteusvacuum commented:
Let me know when you are ready for another review :)

orangejenny committed:
Remove unused fields

orangejenny committed:
total_records variable no longer used

orangejenny committed:
Fewer redraws on pagination

orangejenny commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -0,0 +1,121 @@
+                    self.totalItems(data.total);
+                    self.users.removeAll();     // just in case there are multiple goToPage calls simultaneously
+                    _.each(data.users, function (user) {
+                        self.users.push(mobileWorker(user));
```
154bbef

orangejenny committed:
Naming related to deactivated users

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+    limit = int(request.GET.get('limit', 10))
+    page = int(request.GET.get('page', 1))
+    query = request.GET.get('query')
+    deactivated_only = bool(int(request.GET.get('showDeactivatedUsers', 0)))
```
a00971a

orangejenny committed:
Simplify deactivation logic

orangejenny committed:
Make paginate_mobile_workers location_safe to perhaps fix test

orangejenny commented:
@proteusvacuum Ready! Thank you for the thorough review. There was a bunch of stuff I initially shrugged off, thinking, "I didn't introduce that" but you spending the time to review it motivates me to spend the time to fix it.

orangejenny committed:
Use default paginate_mobile_workers params for test

proteusvacuum commented on corehq/apps/users/static/users/js/mobile_workers.js:
```diff
@@ -78,8 +78,8 @@ hqDefine("users/js/mobile_workers", function () {
-        self.toggleInactiveOnly = function (inactiveOnly) {
-            self.inactiveOnly(inactiveOnly);
+        self.setDeactivatedOnly = function (deactivatedOnly) {
+            self.deactivatedOnly(deactivatedOnly);
```
I still think it's preferable to just set this property directly in the view and set up a listener, that way any future travelers to this code don't have to figure out that they need to use this setter instead. But up to you if you want to keep it this way.

orangejenny replied:
e80ec1b

Also did something similar for `is_active`, replacing the `modifyStatus` with a subscription, because initially I read this comment over email and thought you were talking about `is_active` not `deactivatedOnly`.

orangejenny committed:
Use subscribe more liberally.

w=1 => Had to move the modals so they aren't affected by the visible bindings that control the deactivate &
reactivate buttons.

orangejenny committed:
lint

orangejenny committed:
Use source in ES query

orangejenny commented:
@proteusvacuum Responded to most recent round of comments.

proteusvacuum reviewed (approved):
very nice! :tada:

orangejenny committed:
Add source method to ESQueryFake

orangejenny committed:
Set up _source_fields correctly

orangejenny committed:
Allow chaining of source

orangejenny committed:
Update test_permissions.py

proteusvacuum commented:
:heavy_check_mark: :balloon:
